posydon/popsyn/rate_calculation.py

```diff
@@ -689,20 +703,230 @@ def compute_merger_rate_density(self, w_ijk, z_event, observable='DCOs', sensiti
+            raise ValueError(f'Unknown event {event}!')
+        n = self.df.shape[0]
+        t_birth = self.get_cosmic_time_from_redshift(z_birth) * np.ones(n)  # Gyr
+        t_GRB = t_birth + self.df[f"time_{event}"] * 10 ** (-3) # Gyr
+        return t_GRB
+
+            
+    def compute_GRB_rate_weights(self, sensitivity='infinity', path_to_dir='./', extention='npz'):
+        """Compute the cosmological weights of the transient events associated to the population.
+
+        This function will create a directory path_to_dir/GRBs/sensitivity where
+        it will save the weigths, binary indicies k, z_formation, z_grb.
+        This is needed for scalability as a ~100k DCO population generates ~10M
+        non-zero weights assuming a delta_t=100Myr.
+
+        Parameters
+        ----------
+        sensitivity : string
+            Assume there are no selection effects. Available:
+            'infinite': whole GRB population, i.e. p_det = 1
+        path_to_dir : string
+            Path to the workingn directory where you want to store the
+            cosmological weights.
+
+        """
+        
+        # check if the folder three exists, otherwise create it
+        GRB_dir = os.path.join(path_to_dir,'GRBs')
+        if 'GRBs' not in os.listdir(path_to_dir):
+            os.makedirs(GRB_dir)
+
+        sensitivity_dir = os.path.join(GRB_dir, f'{sensitivity}_sensitivity')
+        if f'{sensitivity}_sensitivity' not in os.listdir(GRB_dir):
+            os.makedirs(sensitivity_dir)
+
+        # index of all DCOs
+        n = self.df.shape[0]
+        index = np.arange(0, n)
+
+        # redshif of each time bin
+        z_birth = self.get_centers_redshift_bins()
+
+        # define interpolator
+        get_redshift_from_time = self.get_redshift_from_cosmic_time_interpolator()
+
+        # hashmap to store everythig
+        data = {'index_1': {}, 'z_grb_1': {}, 'weights_1': {},
+                'index_2': {}, 'z_grb_2': {}, 'weights_2': {}}
+
+        # loop over all redshift bins
+        for i in tqdm(range(len(z_birth))):
+            
+            for s, event in enumerate(['CC1', 'CC2']):
+                # compute the CC time of each compact object
+                t_CC = self.get_time_GRB(z_birth[i], event)
+
+                # intrinsic population
+                if sensitivity == 'infinite':
+                    
+                    # sort out system not emitting GRBs
+                    bool_GRB = np.logical_and(t_CC < cosmology.age(1e-08).value*0.9999999 , self.df[f"GRB{s+1}"])
+                    
+                    # if there are no system emitting any GRB, continue
+                    if len(index[bool_GRB]) == 0:
+                        data[f'index_{s+1}'][str(z_birth[i])] = np.array([])
+                        data[f'z_grb_{s+1}'][str(z_birth[i])] = np.array([])
+                        data[f'weights_{s+1}'][str(z_birth[i])] = np.array([])
+                        continue
```

Comment:
would it be helpful to the user to have a statement here "No system emitting any GRB"?

Reply:
This is part of a large loop, that will execute every delta T ~ 100 Myr so we do not want to spam the user. This information will become clear when plotting rate densities.

Reply:
What you'd usually do in such cases is to define a flag before the loop and set it during the loop, when it happened at least once and make a printout after the loop has finished or only if the flag is still unset.

Reply:
Appending empty arrays is only done for booking when saving and exporting the `index, z_formation, z_grb, w_ijk`. The end product or the user will not care about this dictionary and which redshift bin has empty arrays as they are dropped in the concatenation when exporting the full `index, z_formation, z_grb, w_ijk` arrays.